crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -0,0 +1,100 @@
+use crate::semantic_index::ast_ids::HasScopedAstId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::expression::Expression;
+use crate::semantic_index::symbol::{ScopeId, ScopedSymbolId, SymbolTable};
+use crate::semantic_index::symbol_table;
+use crate::types::{infer_expression_types, IntersectionBuilder, Type, TypeInference};
+use crate::Db;
+use ruff_python_ast as ast;
+use rustc_hash::FxHashMap;
+use std::sync::Arc;
+
+/// Return type constraint, if any, on `definition` applied by `test`.
```

Comment:
I don't understand this comment

Reply:
The expression `test` might (or might not) apply some constraint on `definition`. E.g. if the `definition` is of symbol `x`, then the `test` `x is not None` does apply a constraint on that definition, and we'd return the type `~None` here. But the `test` expression `y is not None` does not apply any constraint to the definition of `x`, so we'd return `None`.

I can add a fuller description/example like this to the doc comment.